examples/src/bin/custom_paintable.rs

```diff
@@ -0,0 +1,100 @@
+//! # Custom GdkPaintable
+//!
+//! This example shows how to create a custom gdk::Paintable
+//! that renders a black square
+
+use gtk::prelude::*;
+use gtk::{gdk, glib, graphene};
+use std::env::args;
+
+mod imp {
+    use super::*;
+    use glib::subclass;
+    use gtk::subclass::prelude::*;
+    pub struct CustomPaintable {}
+
+    impl ObjectSubclass for CustomPaintable {
+        const NAME: &'static str = "CustomPaintable";
+        type Type = super::CustomPaintable;
+        type ParentType = glib::Object;
+        type Instance = subclass::simple::InstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
+
+        glib::object_subclass!();
+
+        fn type_init(type_: &mut subclass::InitializingType<Self>) {
+            type_.add_interface::<gdk::Paintable>();
+        }
+        fn new() -> Self {
+            Self {}
+        }
+    }
+
+    impl ObjectImpl for CustomPaintable {}
+
+    impl PaintableImpl for CustomPaintable {
+        fn get_flags(&self, _paintable: &Self::Type) -> gdk::PaintableFlags {
+            // Fixed size
+            gdk::PaintableFlags::SIZE
+        }
+        fn get_intrinsic_width(&self, _paintable: &Self::Type) -> i32 {
+            200
+        }
+        fn get_intrinsic_height(&self, _paintable: &Self::Type) -> i32 {
+            200
+        }
+        fn snapshot(
+            &self,
+            _paintable: &Self::Type,
+            snapshot: &gdk::Snapshot,
+            width: f64,
+            height: f64,
+        ) {
+            let gtk_snapshot = snapshot.downcast_ref::<gtk::Snapshot>().unwrap();
```

Comment:
I usually just shadow old variables if it's a downcast. You won't need the original one anymore

```suggestion
            let snapshot = snapshot.downcast_ref::<gtk::Snapshot>().unwrap();
```

Reply:
Also it's a bit strange: why do you assume that this is a `gtk::Snapshot`? Is this somehow guaranteed and nothing else can be passed here for some reason, so failure to cast does not have to be handled?

Reply:
I think it's mostly because Snapshot needs to use Gsk types which are not defined in gdk, hence the use of a new type on gtk for that, see https://github.com/aeldemery/gtk4_list_colors/blob/master/sources/ColorWidget.vala#L57 which is a rewrite of a gtk demo in Vala

Reply:
Seems like a strange API choice but whatever :)